### ISSUE TITLE
KeyError on dependency that doesn't exist

### ISSUE THREAD

**Issue by eikenb** [cloudreach/sceptre#603]
```
Traceback (most recent call last):
  File "/home/jae/scratch/venv-work-tools/bin/sceptre", line 11, in <module>
    sys.exit(cli())
  File "/home/jae/scratch/venv-work-tools/lib/python3.5/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/home/jae/scratch/venv-work-tools/lib/python3.5/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/home/jae/scratch/venv-work-tools/lib/python3.5/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/jae/scratch/venv-work-tools/lib/python3.5/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/jae/scratch/venv-work-tools/lib/python3.5/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/jae/scratch/venv-work-tools/lib/python3.5/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/jae/scratch/venv-work-tools/lib/python3.5/site-packages/sceptre/cli/helpers.py", line 37, in decorated
    return func(*args, **kwargs)
  File "/home/jae/scratch/venv-work-tools/lib/python3.5/site-packages/sceptre/cli/launch.py", line 36, in launch_command
    plan = SceptrePlan(context)
  File "/home/jae/scratch/venv-work-tools/lib/python3.5/site-packages/sceptre/plan/plan.py", line 31, in __init__
    all_stacks, command_stacks = config_reader.construct_stacks()
  File "/home/jae/scratch/venv-work-tools/lib/python3.5/site-packages/sceptre/config/reader.py", line 213, in construct_stacks
    stack.dependencies = [stack_map[dep] for dep in stack.dependencies]
  File "/home/jae/scratch/venv-work-tools/lib/python3.5/site-packages/sceptre/config/reader.py", line 213, in <listcomp>
    stack.dependencies = [stack_map[dep] for dep in stack.dependencies]
KeyError: 'csorca/eu-central-1.1/roles.yaml'
```
The config 'csorca/eu-central-1.1/roles.yaml' does not exist and is not referenced. In 'csorca/eu-central-1.1/' there is a 'config.yaml' and a vpc.nodep.yaml file. The latter has no dependencies. In 'csorca/eu-central-1/' there is a different 'config.yaml' and a 'vpc.yaml' and *that* vpc.yaml file does have the dependency `{{'/'.join(command_path)}}/roles.yaml`.

Note that the vpc.yaml, vpc.nodep.yaml and roles.yaml are all sym-links to a common set of stack configs.

Happy to provide more details or help with testing. Pleaes just let me know.

**Comment by eikenb** [comment]
I replaced all uses of sym-links in the config/ area with the files linked to and received the exact same error.

**Comment by eikenb** [comment]
Probably not that  surprising, but if I run with `--ignore-dependencies` it works fine.

**Comment by eikenb** [comment]
If I add a print in the section where that error occurs..
 ```
       for stack in stack_map.values():
            print(stack.name, stack.dependencies)
            if not self.context.ignore_dependencies:
                stack.dependencies = [stack_map[dep] for dep in stack.dependencies]
```
The stack that it errors on is not even the same one. Here's the output from those prints...
```
sceptre launch -y csorca/eu-central-1.1
csorca/eu-central-1.1/vpc.v2.nodep []
csrfodev/us-west-2/roles []
csv2prod/us-west-2/vpc ['csorca/eu-central-1.1/roles.yaml']
Traceback (most recent call last):
```
So it processes one I'm running against and there are no dependences, but then there is one in a later config that has one for here that doesn't exist. Maybe something about using the command_path in the dependencies is causing it?
The dependencies definition, which is in a separate vpc.yaml config, looks like.
```
dependencies:
  - {{'/'.join(command_path)}}/roles.yaml
```

**Comment by ngfgrant** [comment]
Ah ok - yes we promise to handle dependencies strictly on this so it will try and resolve all dependencies for a given stack group or stack unless --ignore-dependencies is set.

**Comment by eikenb** [comment]
It's definitely related to that `command_path`. While it is processing the configurations it uses the path from the command for all of them. So the problem is either that that command_path is not a replacement for the `environment_path` as the migration guide stated (I used the environment_path in v1 for the same purpose) or the other configurations are being processed when the shouldn't be.

So it seems to me that this is one of 2 issues...
1. a bug caused by processing unrelated configs
2. a documentation issue where it states command_path is the same as the old environment_path?

**Comment by eikenb** [comment]
For posterities sake, I just want to mention that command_path is probably set the same as environment_path was...  the reason it is causing issues is that (I think) in v1 only the configurations directly being run were processed, so the path was right. Where in v2 it is processing all the configurations (looking for dependencies) and thus it is wrong for all of them except the 1. This causes my issue as I was using it to get the path relative to the local configuration files.

**Comment by eikenb** [comment]
Aha! Found something that works by printing out the config contents... `stack_group_path` is just what I need. It contains the relative path to the configurations currently being processed. So it can be used to reference the dependencies in that directory dynamically without need for the command_path.

TLDR; switching from `command_path` to `stack_group_path` fixed my issue.

**Comment by stolendog** [comment]
I had a similar issue. It needs to update Migration Guide wiki page. `command_path != environment_path`
Regarding `stack_group_path`, why don't make it as `command_path` which is a list, then we could access necessary items from it like `stack_group_path.0` as we could do it in v1 with `environment_path` If you agree I could make a pull request for it

**Comment by ngfgrant** [comment]
@stolendog iirc we needed to separate out the `command_path` from the `stack_group_path` due to the way we traverse and generate the graph which holds the relationship dependencies and also for the execution of the stacks in that graph. 

I will update the migration guide.

Can i just check, you should definitely be able to reference something like `command_path.0` in your templates as https://github.com/cloudreach/sceptre/commit/62f1323d4ae4d7cfb91a82e58a914f6b030326b9 fixed #561

**Comment by stolendog** [comment]
Definitely, I am able to reference `command_path.0` but it doesn't work for my case, because I need a path relative to the currently processed configuration. If I use command_path.0 for stack_output (`!stack_output {{ command_path.0}}/s3Bucket.yaml::S3BucketName`) in one stack group config, it breaks if I try to work with another stack group because after new v2 dependency model it tries to load all stack groups in single launch despite which stack_group I try to launch.

**Comment by ngfgrant** [comment]
Ok I will update the docs

**Comment by stolendog** [comment]
What about `stack_group_path` issue? should I try to make merge request for `stack_group_path`?

**Comment by stolendog** [comment]
but how can I get folders names from `stack_group_path.0` or `stack_group_path.1` on my configs which are relative to my stack_group?

**Comment by ngfgrant** [comment]
So sorry @stolendog I mis-understood. Ok if you follow the 62f1323 then it should provide some inspiration for that implementation.
